### PR TITLE
 Turn guess_spatial_dimensions to a private function 

### LINKED ISSUE #3396: Deprecate guess_spatial_dimensions
As discussed in #3394, ``color.guess_spatial_dimensions`` is only used in one function ``gaussian.filter``. In addition, this function has a "magics" taste that we decided to limit in the future. #3394 proposed to move the useful code of ``guess_spatial_dimensions`` to the ``_gaussian.py`` module.

### PR DESCRIPTION
**Pull request by sciunto**
## Description

Superseeds #3611
Fixes #3396

## Checklist

- [ ] [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- [ ] Gallery example in `./doc/examples` (new features only)
- [ ] Benchmark in `./benchmarks`, if your changes aren't covered by an
  existing benchmark
- [ ] Unit tests
- [ ] Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)

## For reviewers

- [ ] Check that the PR title is short, concise, and will make sense 1 year
  later.
- [ ] Check that new functions are imported in corresponding `__init__.py`.
- [ ] Check that new features, API changes, and deprecations are mentioned in
      `doc/release/release_dev.rst`.
- [ ] Consider backporting the PR with `@meeseeksdev backport to v0.14.x`

### CODE REVIEW

File: skimage/color/colorconv.py [hunk 1]
```diff
@@ -75,16 +75,9 @@ def guess_spatial_dimensions(image):
...
-    if image.ndim == 3 and image.shape[-1] != 3:
-        return 3
-    if image.ndim == 3 and image.shape[-1] == 3:
-        return None
-    if image.ndim == 4 and image.shape[-1] == 3:
-        return 3
-    else:
-        raise ValueError("Expected 2D, 3D, or 4D array, got %iD." % image.ndim)
+    from ..filters import _guess_spatial_dimensions
+    warn('This function is deprecated and will be removed in 0.18')
```
Comment (hmaarrfk) [inline skimage/color/colorconv.py]: Stack level should be specified

Comment (sciunto) [reply skimage/color/colorconv.py]: Fixed!

**Comment by hmaarrfk** [comment]
I feel like this is used in color conv right? You should use the hidden version of the function there right?

**Comment by hmaarrfk** [comment]
Right.... it isn't in colorconv, which confused me for the longest time. What do you want to do about that? DOn't you think we should be using it in colorconv???

File: skimage/color/colorconv.py [hunk 2]
```diff
@@ -75,16 +75,9 @@ def guess_spatial_dimensions(image):
...
-    if image.ndim == 3 and image.shape[-1] != 3:
-        return 3
-    if image.ndim == 3 and image.shape[-1] == 3:
-        return None
-    if image.ndim == 4 and image.shape[-1] == 3:
-        return 3
-    else:
-        raise ValueError("Expected 2D, 3D, or 4D array, got %iD." % image.ndim)
+    from ..filters import _guess_spatial_dimensions
+    warn('This function is deprecated and will be removed in 0.18', stacklevel=2)
```
Comment (hmaarrfk) [inline skimage/color/colorconv.py]:
```suggestion
    warn('This function is deprecated and will be removed in 0.18', 
         stacklevel=2)
```

**Comment by sciunto** [comment]
Where do you want to use it precisely?

There was a discussion [over here](https://github.com/scikit-image/scikit-image/pull/3394#discussion_r216139334). @jni is for the existence of such functions, but clearly, in the current form, it is not perfect. Likely something to thing more globally for 1.0.

**Comment by hmaarrfk** [comment]
right, i had a PR that did *no guess* and just changed color spaces. It enabled ND color space conversion.

**Comment by sciunto** [comment]
rebased to resolve a conflict.

File: TODO.txt [hunk 3]
```diff
@@ -21,14 +21,14 @@ Version 0.17
 Version 0.18
 ------------
 
+* Remove deprecated function ``guess_spatial_dimensions`` in ``skimage.color.__init__``.
```
Comment (soupault) [inline TODO.txt]: Should it be 0.17?

File: skimage/filters/_gaussian.py [hunk 4]
```diff
@@ -116,3 +115,34 @@ def gaussian(image, sigma=1, output=None, mode='nearest', cval=0,
     image = convert_to_float(image, preserve_range)
     return ndi.gaussian_filter(image, sigma, mode=mode, cval=cval,
                                truncate=truncate)
+
+
+def _guess_spatial_dimensions(image):
```
Comment (soupault) [inline skimage/filters/_gaussian.py]: I think we should take a chance and deprecate this completely from `gaussian`.

File: TODO.txt (hunk 3 above)
Comment (jni) [reply TODO.txt]: No, two versions from the version in which this is released, which will be 0.16.

File: skimage/filters/_gaussian.py (hunk 4 above)
Comment (jni) [reply skimage/filters/_gaussian.py]: Agreed. No more guessing! My apologies for writing that function, as I was young and naive... =)

Comment (stefanv) [reply skimage/filters/_gaussian.py]: What, that's like me apologizing for writing the first version of scikit-image ;)

Comment (jni) [reply skimage/filters/_gaussian.py]: Having said that, I don't think this should hold up this PR, which is already an improvement!

Comment (jni) [reply skimage/filters/_gaussian.py]: @stefanv it would have certainly saved me lots of headaches if you hadn't. =P

Comment (stefanv) [reply skimage/filters/_gaussian.py]: A conversation I have with myself quite frequently :joy:

Comment (sciunto) [reply skimage/filters/_gaussian.py]: :laughing:

**Pull request merged**